appModules/SearchUI: add support for announcing suggestion count in Windows 10 Version 1909 and later

josephsl opened the pull request:
### Link to issue number:
Fixes #10341 

### Summary of the issue:
In Windows 10 Version 1909, File Explorer gains ability to search OneDrive content via Windows Search, with a live region used to provide suggestion count announcement.

### Description of how this pull request fixes the issue:
In Windows 10 Version 1909 (build 18363) and later, File Explorer gains ability to search OneDrive content. This is powered by an embedded Windows Search feature (SearchUI.exe) with a live region announcing suggestion count. This live region has no label, but the child object does, thus coerce the name of this object to be the name of its child, which then allows live region change event to work.

### Testing performed:
Tested in Version 1909 (build 18363) and via Windows 10 App Essentials add-on.

### Known issues with pull request:
None

### Change log entry:
Possibly either new features or changes:

In Windows 10 Version 1909 (change this wording to marketing name once 1909 comes out), NVDA will announce suggestion count when performing searches in File Explorer. (#10231

josephsl committed:
appModules/SearchUI: add support for announcing suggestion count in Windows 10 Version 1909. Re #10341.

In Windows 10 Version 1909 (build 18363) and later, File Explorer gains ability to search OneDrive content. This is powered by an embedded Windows Search feature (SearchUI.exe) with a live region announcing suggestion count. This live region has no label, but the child object does, thus coerce the name of this object to be the name of its child, which then allows live region change event to work.

leonardder commented:
Is this somehow related to #8472?

josephsl commented:
Hi, I can see a connection – let me test that PR to see if this one can be rephrased as such. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, October 23, 2019 1:09 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] appModules/SearchUI: add support for announcing suggestion count in Windows 10 Version 1909 (#10349)

 

Is this somehow related to #8472 <https://github.com/nvaccess/nvda/pull/8472> ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/10349?email_source=notifications&email_token=AB4AXEA4C32TWGDOFCRQJT3QQABCDA5CNFSM4I6NQCOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECAPTLA#issuecomment-545323436> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXECLVJ2WZMWOUKAPZVLQQABCDANCNFSM4I6NQCOA> .

josephsl commented:
Hi, this is different than #8466, as there is no “alert” associated with this one. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Wednesday, October 23, 2019 1:09 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] appModules/SearchUI: add support for announcing suggestion count in Windows 10 Version 1909 (#10349)

 

Is this somehow related to #8472 <https://github.com/nvaccess/nvda/pull/8472> ?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/10349?email_source=notifications&email_token=AB4AXEA4C32TWGDOFCRQJT3QQABCDA5CNFSM4I6NQCOKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECAPTLA#issuecomment-545323436> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXECLVJ2WZMWOUKAPZVLQQABCDANCNFSM4I6NQCOA> .

josephsl commented:
Hi,

It is ready for another round of review after 2019.3 is released. Thanks.

josephsl commented:
Hi,

Indeed, 2019.3 was released.

Thanks.

josephsl commented:
Hi,

Update: I haven't forgotten about this PR - it's just that more important pull requests came along and developers are going through backlog.

Thanks.

josephsl committed:
Merge branch 'master' into i10341searchuiWinTen1909

josephsl committed:
appModules/searchUI: update copyright year.

josephsl committed:
appModules/searchUI: cachedAutomationID -> UIAAutomationId. Re #10341.

josephsl committed:
Merge branch 'master' into i10341searchuiWinTen1909

josephsl committed:
Merge branch 'master' into i10341searchuiWinTen1909

josephsl committed:
appModules/SearchUI: update copyright year

codeofdusk commented:
Any updates on this PR?

josephsl commented:
Hi, I might resurrect this for 2021.2 – see a discussion between I and @JCSTeh for some thoughts about this. In fact, I just committed changes to Windows 10 App Essentials that might help with this PR. Note that announcing suggestion count applies to apps other than SearchUI/SearchApp (Windows Search) because these apps do report suggestion count (navigator object moves to suggestions list which does report position info). Thanks.

josephsl commented:
Hi,
Never mind about the above comment, as I thought it was #7330. The PR is indeed ready to go - I'd wait for 2021.2 cycle begins before requesting another formal review.

Thanks.

josephsl committed:
Merge branch 'master' into i10341searchuiWinTen1909

feerrenrut commented on source/appModules/searchui.py:
```diff
@@ -18,6 +18,15 @@ class StartMenuSearchField(SearchField):
 
+	def event_NVDAObject_init(self, obj):
+		if isinstance(obj, UIA):
+			# #10341: Build 18363 introduces modern search experience in File Explorer.
```
If this is only available in Build 18363 and later, should this be protected by a check for the windows version? If not, please describe in the comment how this code handles earlier versions. I assume that the `UIAAutomationId` wont be `suggestionCountForNarrator` in earlier builds of Windows?

feerrenrut commented on source/appModules/searchui.py:
```diff
@@ -27,7 +36,7 @@ def chooseNVDAObjectOverlayClasses(self,obj,clsList):
 				pass
 		elif isinstance(obj, UIA):
-			if obj.UIAElement.cachedAutomationID == "SearchTextBox":
+			if obj.UIAAutomationId == "SearchTextBox":
```
Perhaps this is only ever true in post 18363 builds?

feerrenrut replied:
Can you verify whether this change is applicable to earlier builds? It may need the same windows version check.

josephsl commented:
Hi, modern search integration was introduced in 18363, and prior to that, the search UI is different (I’ll add a check for windows releases just to make sure). Thanks.

From: Reef Turner ***@***.***> 
Sent: Wednesday, May 19, 2021 3:08 AM
To: nvaccess/nvda ***@***.***>
Cc: Joseph Lee ***@***.***>; Author ***@***.***>
Subject: Re: [nvaccess/nvda] appModules/SearchUI: add support for announcing suggestion count in Windows 10 Version 1909 and later (#10349)

 

@feerrenrut commented on this pull request.

  _____  

In source/appModules/searchui.py <https://github.com/nvaccess/nvda/pull/10349#discussion_r635097543> :

> @@ -18,6 +18,15 @@ class StartMenuSearchField(SearchField):
 
 class AppModule(appModuleHandler.AppModule):
 
+       def event_NVDAObject_init(self, obj):
+              if isinstance(obj, UIA):
+                      # #10341: Build 18363 introduces modern search experience in File Explorer.

If this is only available in Build 18363 and later, should this be protected by a check for the windows version? If not, please describe in the comment how this code handles earlier versions. I assume that the UIAAutomationId wont be suggestionCountForNarrator in earlier builds of Windows?

  _____  

In source/appModules/searchui.py <https://github.com/nvaccess/nvda/pull/10349#discussion_r635097864> :

> @@ -27,7 +36,7 @@ def chooseNVDAObjectOverlayClasses(self,obj,clsList):
                       except ValueError:
                               pass
               elif isinstance(obj, UIA):
-                      if obj.UIAElement.cachedAutomationID == "SearchTextBox":
+                      if obj.UIAAutomationId == "SearchTextBox":

Perhaps this is only ever true in post 18363 builds?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/10349#pullrequestreview-663005949> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEECGSJEPBH4NO4TMVLTOOEZ3ANCNFSM4I6NQCOA> .

feerrenrut commented:
Ok. But the problem I want to solve is:
"From the perspective of a code reader, how do they know that this code will only run on the intended versions of windows?"

josephsl commented:
Hi, one simple solution I think (which I will do before I sleep) is specifying on which Windows 10 version the code will run on (winVersion.WIN10_1909, to be exact). Thanks.

josephsl committed:
Merge branch 'master' into i10341searchuiWinTen1909

josephsl committed:
SearchUI: specify that suggestion count from Windows Search is available from Windows 10 Version 1909 (build 18363/NOvember 2019 Update). Re #10341.

Comment from Reef Turner (NV Access): clarify that Windows Search/suggestion count in File Explorer is available from Windows 10 Version 1909, also split the condition into multiple lines to prepare for Flake8 (lint).

josephsl commented:
Hi, no need to change this, as SearchUI was introduced in Windows 10, although I think a header comment should clarify this fact (@MichaelDCurran had an adventure with this app module in particular back in 2015 because Windows Search kept freezing for some odd reason). Thanks,

josephsl commented:
Hi,

A component of #12585 because searchapp.exe was renamed to searchhost.exe in Windows 11, complete with exact same live region implementation as this PR. Therefore, once this PR is erviewed and merged, it can be applied easily to Windows 11 by introducing an alias app module for searchhost.exe that will import from searchui.py.

Thanks.

lukaszgo1 commented:
@feerrenrut  It looks like all changes that you've requested has been addressed. Would it be possible to re-review this?

feerrenrut committed:
Merge remote-tracking branch 'origin/master' into i10341searchuiWinTen1909

feerrenrut commented:
How to test: 
- Open explorer window `Windows + e`
- Press F3 to focus search
- Type something. Note "10 suggestions available, press down arrow key to access results" announced, this is new.
- Press down arrow, file name and details should be announced

feerrenrut committed:
Update changes file for PR #10349

feerrenrut reviewed (approved):
Thanks @josephsl
